Add support for named parameters in the API

dacort opened the pull request:
Currently, extensions set named parameters directly on the TableFunction object. For rust-based extensions, I'm not sure this is easily doable, so this PR adds two new APIs `duckdb_table_function_add_named_parameter` and `duckdb_bind_get_named_parameter`.

dacort committed:
Add support for named paramters in the API

Mytherin commented on test/api/capi/capi_table_functions.cpp:
```diff
@@ -65,6 +65,7 @@ static void capi_register_table_function(duckdb_connection connection, const cha
 	// add a string parameter
 	duckdb_logical_type type = duckdb_create_logical_type(DUCKDB_TYPE_BIGINT);
 	duckdb_table_function_add_parameter(function, type);
+	duckdb_table_function_add_named_parameter(function, "my_parameter", type);
```
Could you add a test where the named parameter is used in a SQL query?

dacort replied:
Will do!

dacort replied:
Updated with an additional test.

Mytherin commented on src/include/duckdb.h:
```diff
@@ -1710,6 +1710,16 @@ Adds a parameter to the table function.
+* name: The name of the parameter
+* type: The type of the parameter to add.
+*/
+DUCKDB_API void duckdb_table_function_add_named_parameter(duckdb_table_function table_function, const char *name,
```
Should there not also be a corresponding `duckdb_bind_get_named_parameter` or similar to allow access to these parameters during the bind phase?

dacort replied:
Yes, you're right. Have that in a local branch and will push up.

dacort replied:
Added `duckdb_bind_get_named_parameter`

Mytherin reviewed (commented):
Thanks for the PR! LGTM. Some comments:

dacort commented:
@Mytherin Addressed the comments and all the tests except for the "Python workflow run" succeeded. I'm unsure if that's related - the run [timed out after 6 hours](https://github.com/duckdb/duckdb/actions/runs/4338489074/jobs/7575221063). 

Squashed my changes and running through the test suite again now.

Tishj commented:
Yea that issue was mentioned, not related to this.
We're still investigating the cause of that problem :)

Mytherin commented on test/api/capi/capi_table_functions.cpp:
```diff
@@ -93,6 +98,10 @@ TEST_CASE("Test Table Functions C API", "[capi]") {
 	REQUIRE_NO_FAIL(*result);
 	REQUIRE(result->Fetch<int64_t>(0, 0) == 42);
 
+	result = tester.Query("SELECT * FROM my_function(1, my_parameter=3)");
```
Can we add some tests for the various edge cases as well (e.g. fetching a parameter that isn't there, pushing `nullptr` into the various arguments, etc)?

dacort replied:
Sure thing.

dacort replied:
@Mytherin Is there any preference for extending the original tests vs. creating a completely new test case? For the additional tests, I would need to update the existing bind/init/function definitions. It shouldn't be a problem, but not sure if we want to just incorporate the new API functionality or leave the existing test cases as is.

Mytherin replied:
Perhaps you could create a new test case for this?

dacort replied:
Sounds good, thanks for the quick reply!

dacort committed:
Add some additional tests

dacort committed:
format fix

Mytherin commented:
Thanks! Looks good to me.
